### PR TITLE
OXDEV-3483 OXDEV-3792 Fix dependencies

### PR DESCRIPTION
**Pull request by hkreuter**
PR is only to make it easier to get some feedback

### CODE REVIEW

File: src/Account/Infrastructure/CustomerRegisterFactory.php [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    private $authentication;
+
+    /** @var Repository */
+    private $repository;
+
+    /** @var Legacy */
+    private $legacyService;
+
+    public function __construct(
+        Authentication $authentication,
```
Comment (realFlowControl) [inline src/Account/Infrastructure/CustomerRegisterFactory.php]: Is not used and should not be as infra is not allowed to depend on services

File: src/Account/Infrastructure/CustomerRegisterFactory.php [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+    public function createValidCustomer(string $email, string $password, ?DateTimeInterface $birthdate): Customer
+    {
+        if (!strlen($email)) {
+            throw new EmailEmpty();
+        }
+
+        if (!$this->legacyService->isValidEmail($email)) {
+            throw new InvalidEmail();
+        }
```
Comment (realFlowControl) [inline src/Account/Infrastructure/CustomerRegisterFactory.php]: I would prefer to have on `InvalidEmail` exception only and give two different reasons by named constructor.

`InvalidEmail::byEmptyString` / `InvalidEmail::byString($email)` for example

File: src/Account/Infrastructure/CustomerRegisterFactory.php [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+        if (strlen($password) == 0 ||
+            (strlen($password) < $this->legacyService->getConfigParam('iPasswordLength'))
+        ) {
```
Comment (realFlowControl) [inline src/Account/Infrastructure/CustomerRegisterFactory.php]: could you check if there is a public service in core that validates a password?

File: src/Account/Infrastructure/CustomerRegisterFactory.php [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+            throw new Password();
+        }
+
+        if ($this->repository->checkEmailExists($email)) {
+            throw CustomerExists::byEmail($email);
+        }
```
Comment (realFlowControl) [inline src/Account/Infrastructure/CustomerRegisterFactory.php]: this validation part can stay in the service, as this is business logic and should not go to the infra layer

File: src/Account/Service/CustomerRegisterInput.php [hunk 5]
```diff
@@ -10,75 +10,26 @@
...
-
-        if (strlen($password) == 0 ||
-            (strlen($password) < $this->legacyService->getConfigParam('iPasswordLength'))
-        ) {
-            throw new Password();
-        }
-
-        if ($this->repository->checkEmailExists($email)) {
-            throw CustomerExists::byEmail($email);
-        }
```
Comment (realFlowControl) [inline src/Account/Service/CustomerRegisterInput.php]: as stated above, this should stay and does not belong to infra layer

File: src/NewsletterStatus/Service/NewsletterStatus.php [hunk 6]
```diff
@@ -14,6 +14,7 @@
 use OxidEsales\GraphQL\Account\NewsletterStatus\DataType\NewsletterStatusUnsubscribe as NewsletterStatusUnsubscribeType;
 use OxidEsales\GraphQL\Account\NewsletterStatus\DataType\Subscriber as SubscriberType;
 use OxidEsales\GraphQL\Account\NewsletterStatus\Exception\SubscriberNotFound;
+use OxidEsales\GraphQL\Account\NewsletterStatus\Infrastructure\NewsletterStatus as NewsletterStatusInfrastructure;
 use OxidEsales\GraphQL\Account\NewsletterStatus\Infrastructure\Repository as NewsletterStatusRepository;
```
Comment (realFlowControl) [inline src/NewsletterStatus/Service/NewsletterStatus.php]: we should rethink the naming, `NewsletterStatusInfrastructure` seems misleading, as it is a Repository. I would suggest to rename the Repostory to just `NewsletterRepository` or maybe better `NewsletterSubscriptionRepository` and then rename `NewsletterStatusInfrastructure` to `NewsletterStatusRepository`

File: src/WishedPrice/Infrastructure/WishedPriceFactory.php [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+        try {
+            /** @var ProductDataType $product */
+            $product = $this->repository->getById($id, ProductDataType::class);
+        } catch (NotFound $e) {
+            throw ProductNotFound::byId($id);
+        }
+
+        // Throw 404 if product has wished prices disabled
+        if (!$product->getEshopModel()->isPriceAlarm()) {
+            throw ProductNotFound::byId($id);
```
Comment (realFlowControl) [inline src/WishedPrice/Infrastructure/WishedPriceFactory.php]: Just an opinion: a `404` status code seems wrong here, but I don't know of a better matching status code then the overly generic `400` ..

File: src/WishedPrice/Service/WishedPriceInput.php [hunk 8]
```diff
@@ -26,82 +20,28 @@ final class WishedPriceInput
...
+        $this->wishedPriceFactory = $wishedPriceFactory;
     }
 
     /**
      * @Factory
      */
     public function fromUserInput(ID $productId, string $currencyName, float $price): WishedPrice
     {
-        $this->assertProductWishedPriceIsPossible($productId);
-        $this->assertPriceValue($price);
```
Comment (realFlowControl) [inline src/WishedPrice/Service/WishedPriceInput.php]: Validation is ok here, should not go to Infra layer

**Review by realFlowControl — CHANGES_REQUESTED** [review]

File: src/Account/Infrastructure/CustomerRegisterFactory.php (hunk 3 above)
Comment (hkreuter) [reply src/Account/Infrastructure/CustomerRegisterFactory.php]: From what I've seen no, but we can probably use the shop's  InputValidator.

File: src/WishedPrice/Infrastructure/WishedPriceFactory.php (hunk 7 above)
Comment (hkreuter) [reply src/WishedPrice/Infrastructure/WishedPriceFactory.php]: maybe use 418 I'm a teapot ?  or 405 Method Not Allowed ?

File: src/Account/Infrastructure/CustomerRegisterFactory.php (hunk 3 above)
Comment (hkreuter) [reply src/Account/Infrastructure/CustomerRegisterFactory.php]: Had a closer look, and no, from what I've seen, shop does not offer anything that fits our needs atm.

File: src/WishedPrice/Infrastructure/WishedPriceFactory.php (hunk 7 above)
Comment (realFlowControl) [reply src/WishedPrice/Infrastructure/WishedPriceFactory.php]: I guess we can just leave the `404`

Comment (hkreuter) [reply src/WishedPrice/Infrastructure/WishedPriceFactory.php]: Point is: 404 give the hint that product is not available for wished price. 400 would only be some 'bad request'. Looks like we'll leave it at 404 for now.

**Review by realFlowControl — APPROVED** [review]

**Comment by hkreuter** [comment]
Merge is done in OXDEV-3792.

**Pull request closed without merging**